--- a/target/linux/airoha/dts/an7581.dtsi
+++ b/target/linux/airoha/dts/an7581.dtsi
@@ -156,78 +156,78 @@
 		compatible = "operating-points-v2";
 		opp-shared;
 
-		opp-500000000 {
-			opp-hz = /bits/ 64 <500000000>;
-			required-opps = <&smcc_opp0>;
-		};
-
-		opp-550000000 {
-			opp-hz = /bits/ 64 <550000000>;
-			required-opps = <&smcc_opp1>;
-		};
-
-		opp-600000000 {
-			opp-hz = /bits/ 64 <600000000>;
-			required-opps = <&smcc_opp2>;
-		};
-
-		opp-650000000 {
-			opp-hz = /bits/ 64 <650000000>;
-			required-opps = <&smcc_opp3>;
-		};
-
 		opp-700000000 {
 			opp-hz = /bits/ 64 <700000000>;
-			required-opps = <&smcc_opp4>;
+			required-opps = <&smcc_opp0>;
 		};
 
 		opp-750000000 {
 			opp-hz = /bits/ 64 <750000000>;
-			required-opps = <&smcc_opp5>;
+			required-opps = <&smcc_opp1>;
 		};
 
 		opp-800000000 {
 			opp-hz = /bits/ 64 <800000000>;
-			required-opps = <&smcc_opp6>;
+			required-opps = <&smcc_opp2>;
 		};
 
 		opp-850000000 {
 			opp-hz = /bits/ 64 <850000000>;
-			required-opps = <&smcc_opp7>;
+			required-opps = <&smcc_opp3>;
 		};
 
 		opp-900000000 {
 			opp-hz = /bits/ 64 <900000000>;
-			required-opps = <&smcc_opp8>;
+			required-opps = <&smcc_opp4>;
 		};
 
 		opp-950000000 {
 			opp-hz = /bits/ 64 <950000000>;
-			required-opps = <&smcc_opp9>;
+			required-opps = <&smcc_opp5>;
 		};
 
 		opp-1000000000 {
 			opp-hz = /bits/ 64 <1000000000>;
-			required-opps = <&smcc_opp10>;
+			required-opps = <&smcc_opp6>;
 		};
 
 		opp-1050000000 {
 			opp-hz = /bits/ 64 <1050000000>;
-			required-opps = <&smcc_opp11>;
+			required-opps = <&smcc_opp7>;
 		};
 
 		opp-1100000000 {
 			opp-hz = /bits/ 64 <1100000000>;
-			required-opps = <&smcc_opp12>;
+			required-opps = <&smcc_opp8>;
 		};
 
 		opp-1150000000 {
 			opp-hz = /bits/ 64 <1150000000>;
-			required-opps = <&smcc_opp13>;
+			required-opps = <&smcc_opp9>;
 		};
 
 		opp-1200000000 {
 			opp-hz = /bits/ 64 <1200000000>;
+			required-opps = <&smcc_opp10>;
+		};
+
+		opp-1250000000 {
+			opp-hz = /bits/ 64 <1250000000>;
+			required-opps = <&smcc_opp11>;
+		};
+
+		opp-1300000000 {
+			opp-hz = /bits/ 64 <1300000000>;
+			required-opps = <&smcc_opp12>;
+		};
+
+		opp-1350000000 {
+			opp-hz = /bits/ 64 <1350000000>;
+			required-opps = <&smcc_opp13>;
+		};
+
+		opp-1400000000 {
+			opp-hz = /bits/ 64 <1400000000>;
 			required-opps = <&smcc_opp14>;
 		};
 	};
