--- a/target/linux/airoha/dts/an7581.dtsi
+++ b/target/linux/airoha/dts/an7581.dtsi
@@ -30,6 +30,14 @@
 			reg = <0x0 0x84000000 0x0 0xa00000>;
 		};
 
+		/* 64 KiB reserved for ramoops/pstore */
+		ramoops@86ff0000 {
+			no-map;
+			compatible = "ramoops";
+			reg = <0x0 0x86ff0000 0x0 0x10000>;
+			record-size = <0x1000>;
+		};
+
 		qdma0_buf: qdma0-buf@87000000 {
 			no-map;
 			reg = <0x0 0x87000000 0x0 0x2000000>;
